src/coordinator/functions/temporal/base.go

```diff
@@ -0,0 +1,339 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package temporal
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3db/src/coordinator/block"
+	"github.com/m3db/m3db/src/coordinator/executor/transform"
+	"github.com/m3db/m3db/src/coordinator/parser"
+)
+
+var emptyOp = baseOp{}
+
+// baseOp stores required properties for logical operations
+type baseOp struct {
+	operatorType string
+	duration     time.Duration
+	processorFn  MakeProcessor
+}
+
+func newBaseOp(args []interface{}, operatorType string, processorFn MakeProcessor) (baseOp, error) {
+	if len(args) != 1 {
+		return emptyOp, fmt.Errorf("invalid number of args for %s: %d", operatorType, len(args))
+	}
+
+	duration, ok := args[0].(time.Duration)
+	if !ok {
+		return emptyOp, fmt.Errorf("unable to cast to scalar argument: %v", args[0])
+	}
+
+	return baseOp{
+		operatorType: operatorType,
+		processorFn:  processorFn,
+		duration:     duration,
+	}, nil
+}
+
+// OpType for the operator
+func (o baseOp) OpType() string {
+	return o.operatorType
+}
+
+// String representation
+func (o baseOp) String() string {
+	return fmt.Sprintf("type: %s, duration: %v", o.OpType(), o.duration)
+}
+
+// Node creates an execution node
+func (o baseOp) Node(controller *transform.Controller, opts transform.Options) transform.OpNode {
+	return &baseNode{
+		controller:    controller,
+		cache:         transform.NewTimeCache(),
+		op:            o,
+		processor:     o.processorFn(o, controller),
+		transformOpts: opts,
+	}
+}
+
+// baseNode is an execution node
+type baseNode struct {
+	op            baseOp
+	controller    *transform.Controller
+	cache         *transform.TimeCache
+	processor     Processor
+	mu            sync.Mutex
+	transformOpts transform.Options
+}
+
+// Process processes a block. The processing steps are as follows:
+// 1. Figure out the maximum blocks needed for the temporal function
+// 2. For the current block, figure out whether we have enough previous blocks which can help process it
+// 3. For the blocks after current block, figure out which can be processed right now
+// 4. Process all valid blocks from #3, #4 and mark them as processed
+// 5. Run a sweep face to free up blocks which are no longer needed to be cached
+func (c *baseNode) Process(ID parser.NodeID, b block.Block) error {
```

Comment:
Could you add some comments in the code explaining what each section is doing? I think that'd be very helpful.